Feature/stale while revalidate

tijlleenders opened the pull request:
@GithinjiHans @Tushar-4781 

Gave the offline-first (stale-while-revalidate) caching for offline mode a go.
TLDR: Local is working as well as the manually deployed zinzen.me :)

![image](https://user-images.githubusercontent.com/44887755/173222701-325e1420-a22f-49e6-a408-1c2d860aeaeb.png)


Will need some help to make this PR clean and configure (pre-)caching details and all - but I think the basics are here.

I used the create-react-app PWA template as a basis for the registerSW and service-worker files.
This is I believe also the basis @mallikarai05 used for initiating this codebase.

I first tried withoug the Vite-PWA plugin...
The build/bundle process makes things more complicated as now we have to instruct it to also copy the service worker file - but not add a hash to it? Or maybe we do want the hash for that too? Anyway - I couldn't configure it right even though it is quite [well documented](https://rollupjs.org/guide/en/). 
This is the downside of using a builder/bundler: it creates another tool to learn and you can't easily relate source files to deployed files.
I left the rollup code commented out. If it is glaringly obvious what I'm doing wrong please teach me :)
![image](https://user-images.githubusercontent.com/44887755/173222366-fb871183-76b6-4c5a-bfc2-9ecc4466f156.png)

Then I tried using the Vite-PWA plugin which nearly worked immediately.

So we now need to choose to go the VitePWA route or stick with Rollup to avoid yet another dev dependency/plugin.

pulkitvyas08 reviewed (changes requested):
VitePWA seems to be a good choice. @Tushar-4781 can you look into fixing the cypress errors for this PR and maybe extend this work to add offline installation enabled for desktop, even if that doesn't get merged with this PR we will get an idea if this setup is useful or not

tijlleenders commented:
Good idea.

I just checked for mobile and there it seems to be installing. We might want to make a separate issue to catch the install event so we can couple it with a button in the menu.

![Screenshot_20220612-172859_Chrome.jpg](https://user-images.githubusercontent.com/44887755/173240671-e3d23db4-1077-443b-9c3d-8d7ad1cf0650.jpg)

![Screenshot_20220612-172912_Chrome.jpg](https://user-images.githubusercontent.com/44887755/173240687-47fc3cd4-d72f-4058-be61-89a59e085e86.jpg)

![Screenshot_20220612-172934_One UI Home.jpg](https://user-images.githubusercontent.com/44887755/173240700-cf304fc8-db80-4a09-9f42-23259204b5a2.jpg)

pulkitvyas08 commented:
> 

But you are testing on [zinzen.me](https://zinzen.me/). For this PR's preview URL I'm getting this error `Error during service worker registration: TypeError: ServiceWorker script at https://zinzen-git-feature-stale-while-revalidate-tijlleenders.vercel.app/service-worker.js for scope https://zinzen-git-feature-stale-while-revalidate-tijlleenders.vercel.app/ threw an exception during script evaluation.`

tijlleenders commented:
There is something I removed that might be able to solve this. 

The two base files you get when using create React app (CRA) with template PWA for Typescript mention a public URL env variable - but I removed that since it was undefined when running on localhost.

tijlleenders commented:
Oh wait - they're identical already.
Hmmm... that could mean trouble.

tijlleenders commented:
@pulkitvyas08 I get the same error when building on localhost:
```
Uncaught non-precached-url: non-precached-url :: [{"url":"undefined/index.html"}]
```

I had solved it by adding this to the Vite-PWA plugin - but forgot to commit the fix:
```
      workbox: {
        cleanupOutdatedCaches: true, //default is true but added to make it explicit
        ignoreURLParametersMatching: []
      },
```
Committed it now.

![image](https://user-images.githubusercontent.com/44887755/173244439-b5f81b7a-798a-4ca9-ba13-2e5a837b27a1.png)

Tushar-4781 commented:
@tijlleenders where is the **manifest** file

tijlleenders commented:
> @tijlleenders where is the **manifest** file

It is being generated by the Vite-PWA plugin (hardcoded in the config - could be moved out if you prefer).
I removed the separate file because we should only have the config in one place.

tijlleenders commented:
![image](https://user-images.githubusercontent.com/44887755/173629867-e804e07c-e265-48db-bb4d-3656d89be306.png)

Works on Vercel now too. It was missing the icons.

@Tushar-4781 Could you finish up the test case fix?
Then I think we can merge.

Tushar-4781 reviewed (approved):
this will be the **first biggest achievement** in the **history of the ZinZen App** Cheers 🍻

pulkitvyas08 reviewed (approved):
This is amazing work! Installation doesn't work for firefox but does for chromium based browsers so good enough. Approved and merging 🚀

tijlleenders commented:
Closes #231 #207 #208
